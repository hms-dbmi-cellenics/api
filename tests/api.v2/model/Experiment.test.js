// Disabled ts because it doesn't recognize jest mocks
// @ts-nocheck
const sqlClient = require('../../../src/sql/sqlClient');
const helpers = require('../../../src/sql/helpers');

const validSamplesOrderResult = ['sampleId1', 'sampleId2', 'sampleId3', 'sampleId4'];
const getProcessingConfigResponse = require('../mocks/data/getProcessingConfigResponse.json');
const { mockSqlClient, mockTrx } = require('../mocks/getMockSqlClient')();
const BasicModel = require('../../../src/api.v2/model/BasicModel');
const { mockS3GetSignedUrl } = require('../../test-utils/mockAWSServices');
const config = require('../../../src/config');

jest.mock('../../../src/sql/sqlClient', () => ({
  get: jest.fn(() => mockSqlClient),
}));

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'mockNewExperimentId'),
}));

jest.mock('../../../src/sql/helpers', () => ({
  collapseKeysIntoObject: jest.fn(),
  collapseKeyIntoArray: jest.fn(),
  replaceNullsWithObject: () => (`COALESCE(
      jsonb_object_agg(pipeline_type, jsonb_build_object('state_machine_arn', state_machine_arn, 'execution_arn', execution_arn))
      FILTER(
        WHERE pipeline_type IS NOT NULL
      ),
      '{}'::jsonb
    ) as pipelines,}),`),
}));

const Experiment = require('../../../src/api.v2/model/Experiment');
const constants = require('../../../src/utils/constants');
const tableNames = require('../../../src/api.v2/model/tableNames');

const mockExperimentId = 'mockExperimentId';
const mockSampleId = 'mockSampleId';
const mockExperimentName = 'mockNewName';

describe('model/Experiment', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('getAllExperiments works correctly', async () => {
    const queryResult = 'result';
    helpers.collapseKeyIntoArray.mockReturnValueOnce(
      Promise.resolve(queryResult),
    );

    const expectedResult = await new Experiment().getAllExperiments('mockUserId');

    expect(queryResult).toEqual(expectedResult);

    expect(sqlClient.get).toHaveBeenCalled();
    expect(helpers.collapseKeyIntoArray.mock.calls).toMatchSnapshot();

    expect(mockSqlClient.select).toHaveBeenCalledWith(
      [
        'e.id',
        'e.name',
        'e.description',
        'e.samples_order',
        'e.notify_by_email',
        'e.pipeline_version',
        'e.created_at',
        'e.updated_at',
        'm.key',
        'p.parent_experiment_id',
        mockSqlClient.raw('CASE WHEN p.experiment_id IS NOT NULL THEN true ELSE false END as is_subsetted'),
      ],
    );
    expect(mockSqlClient.from).toHaveBeenCalledWith('user_access');
    expect(mockSqlClient.where).toHaveBeenCalledWith('user_id', 'mockUserId');
    expect(mockSqlClient.join).toHaveBeenCalledWith('experiment as e', 'e.id', 'user_access.experiment_id');
    expect(mockSqlClient.leftJoin).toHaveBeenCalledWith('metadata_track as m', 'e.id', 'm.experiment_id');
    expect(mockSqlClient.as).toHaveBeenCalledWith('mainQuery');
  });

  it('getExampleExperiments works correctly', async () => {
    const queryResult = 'result';

    mockSqlClient.groupBy.mockReturnValueOnce(queryResult);

    const expectedResult = await new Experiment().getExampleExperiments();

    expect(queryResult).toEqual(expectedResult);

    expect(sqlClient.get).toHaveBeenCalled();
    expect(mockSqlClient.select).toHaveBeenCalledWith(
      [
        'e.id',
        'e.name',
        'e.description',
        'e.publication_title',
        'e.publication_url',
        'e.data_source_title',
        'e.data_source_url',
        'e.species',
        'e.cell_count',
      ],
    );
    expect(mockSqlClient.min).toHaveBeenCalledWith('s.sample_technology as sample_technology');
    expect(mockSqlClient.count).toHaveBeenCalledWith('s.id as sample_count');
    expect(mockSqlClient.from).toHaveBeenCalledWith(tableNames.USER_ACCESS);
    expect(mockSqlClient.join).toHaveBeenCalledWith(`${tableNames.EXPERIMENT} as e`, 'e.id', `${tableNames.USER_ACCESS}.experiment_id`);
    expect(mockSqlClient.join).toHaveBeenCalledWith(`${tableNames.SAMPLE} as s`, 'e.id', 's.experiment_id');
    expect(mockSqlClient.where).toHaveBeenCalledWith('user_id', constants.PUBLIC_ACCESS_ID);
    expect(mockSqlClient.groupBy).toHaveBeenCalledWith('e.id');
  });

  it('getExperimentData works correctly', async () => {
    const experimentFields = [
      'id', 'name', 'description',
      'samples_order', 'processing_config',
      'notify_by_email', 'pipeline_version',
      'created_at', 'updated_at',
    ];

    const queryResult = 'result';
    helpers.collapseKeysIntoObject.mockReturnValueOnce(mockSqlClient);
    mockSqlClient.first.mockReturnValueOnce(queryResult);

    const mockCollapsedObject = 'collapsedObject';
    mockSqlClient.raw.mockImplementationOnce(() => mockCollapsedObject);

    const expectedResult = await new Experiment().getExperimentData(mockExperimentId);

    expect(expectedResult).toEqual(queryResult);

    expect(sqlClient.get).toHaveBeenCalled();

    expect(mockSqlClient.raw.mock.calls[0]).toMatchSnapshot();

    expect(mockSqlClient.select).toHaveBeenCalledWith([...experimentFields, 'parent_experiment_id', mockCollapsedObject]);
    expect(mockSqlClient.groupBy).toHaveBeenCalledWith([...experimentFields, 'parent_experiment_id']);
    expect(mockSqlClient.from).toHaveBeenCalled();

    // Check that mainQuery is correct
    const mainQuery = mockSqlClient.from.mock.calls[0][0];

    jest.clearAllMocks();
    await mainQuery.bind(mockSqlClient)();

    expect(mockSqlClient.select).toHaveBeenCalledWith('*');
    expect(mockSqlClient.from).toHaveBeenCalledWith('experiment');
    expect(mockSqlClient.leftJoin).toHaveBeenCalledWith('experiment_execution', 'experiment.id', 'experiment_execution.experiment_id');
    expect(mockSqlClient.where).toHaveBeenCalledWith('id', 'mockExperimentId');
    expect(mockSqlClient.as).toHaveBeenCalledWith('mainQuery');
  });

  it('getExperimentData throws if an empty object is returned (the experiment was not found)', async () => {
    helpers.collapseKeysIntoObject.mockReturnValueOnce(mockSqlClient);
    mockSqlClient.first.mockReturnValueOnce({});

    await expect(new Experiment().getExperimentData(mockExperimentId)).rejects.toThrow(new Error('Experiment not found'));
  });

  it('createCopy works correctly', async () => {
    mockSqlClient.raw.mockImplementation((template, values) => {
      if (values && values.length) return template.replace('?', values[0]);
      return template;
    });

    mockSqlClient.where.mockImplementationOnce(() => 'mockQuery');

    const expectedResult = await new Experiment().createCopy(mockExperimentId, mockExperimentName);

    expect(expectedResult).toEqual('mockNewExperimentId');

    expect(sqlClient.get).toHaveBeenCalled();

    expect(mockSqlClient.insert).toHaveBeenCalledWith('mockQuery');

    expect(mockSqlClient.select).toHaveBeenCalledWith([
      'mockNewExperimentId as id',
      'mockNewName as name',
      'description',
      'pod_cpus',
      'pod_memory',
    ]);

    expect(mockSqlClient.where).toHaveBeenCalledWith({ id: mockExperimentId });
    expect(mockSqlClient.into).toHaveBeenCalledWith('experiment (id, name, description, pod_cpus, pod_memory)');
  });

  it('createCopy works correctly without a name', async () => {
    mockSqlClient.raw.mockImplementation((template, values) => {
      if (values && values.length) return template.replace('?', values[0]);
      return template;
    });

    mockSqlClient.where.mockImplementationOnce(() => 'mockQuery');

    const expectedResult = await new Experiment().createCopy(mockExperimentId);

    expect(expectedResult).toEqual('mockNewExperimentId');

    expect(sqlClient.get).toHaveBeenCalled();

    expect(mockSqlClient.insert).toHaveBeenCalledWith('mockQuery');

    expect(mockSqlClient.select).toHaveBeenCalledWith([
      'mockNewExperimentId as id',
      'name',
      'description',
      'pod_cpus',
      'pod_memory',
    ]);

    expect(mockSqlClient.where).toHaveBeenCalledWith({ id: mockExperimentId });
    expect(mockSqlClient.into).toHaveBeenCalledWith('experiment (id, name, description, pod_cpus, pod_memory)');
  });

  it('updateSamplePosition works correctly if valid params are passed', async () => {
    mockTrx.returning.mockImplementationOnce(
      () => Promise.resolve([{ samplesOrder: validSamplesOrderResult }]),
    );
    mockTrx.raw.mockImplementationOnce(() => 'resultOfSql.raw');

    await new Experiment().updateSamplePosition(mockExperimentId, 0, 1);

    expect(mockTrx).toHaveBeenCalledWith('experiment');

    expect(mockTrx.update).toHaveBeenCalledWith({ samples_order: 'resultOfSql.raw' });
    expect(mockTrx.raw.mock.calls[0]).toMatchSnapshot();
    expect(mockTrx.where).toHaveBeenCalledWith('id', 'mockExperimentId');
    expect(mockTrx.returning).toHaveBeenCalledWith(['samples_order']);

    expect(mockTrx.commit).toHaveBeenCalled();
    expect(mockTrx.rollback).not.toHaveBeenCalled();
  });

  it('updateSamplePosition rolls back if the result is invalid', async () => {
    mockTrx.returning.mockImplementationOnce(() => Promise.resolve([{ samplesOrder: null }]));
    mockTrx.raw.mockImplementationOnce(() => 'resultOfSql.raw');

    await expect(new Experiment().updateSamplePosition(mockExperimentId, 0, 1)).rejects.toThrow('Invalid update parameters');

    expect(mockTrx).toHaveBeenCalledWith('experiment');

    expect(mockTrx.update).toHaveBeenCalledWith({ samples_order: 'resultOfSql.raw' });
    expect(mockTrx.raw.mock.calls[0]).toMatchSnapshot();
    expect(mockTrx.where).toHaveBeenCalledWith('id', 'mockExperimentId');
    expect(mockTrx.returning).toHaveBeenCalledWith(['samples_order']);

    expect(mockTrx.commit).not.toHaveBeenCalled();
    expect(mockTrx.rollback).toHaveBeenCalled();
  });

  it('updateSamplePosition rolls back if the parameters are invalid', async () => {
    mockTrx.returning.mockImplementationOnce(
      () => Promise.resolve([{ samplesOrder: validSamplesOrderResult }]),
    );
    mockTrx.raw.mockImplementationOnce(() => 'resultOfSql.raw');

    await expect(new Experiment().updateSamplePosition(mockExperimentId, 0, 10000)).rejects.toThrow('Invalid update parameters');

    expect(mockTrx).toHaveBeenCalledWith('experiment');

    expect(mockTrx.update).toHaveBeenCalledWith({ samples_order: 'resultOfSql.raw' });
    expect(mockTrx.raw.mock.calls[0]).toMatchSnapshot();
    expect(mockTrx.where).toHaveBeenCalledWith('id', 'mockExperimentId');
    expect(mockTrx.returning).toHaveBeenCalledWith(['samples_order']);

    expect(mockTrx.commit).not.toHaveBeenCalled();
    expect(mockTrx.rollback).toHaveBeenCalled();
  });

  it('addSamples works correctly', async () => {
    mockSqlClient.where.mockImplementationOnce(() => { Promise.resolve(); });
    mockSqlClient.raw.mockImplementationOnce(() => 'RawSqlCommand');

    const newSamples = [mockSampleId, 'id2'];

    await new Experiment().addSamples(mockExperimentId, newSamples);

    expect(mockSqlClient.update).toHaveBeenCalledWith({ samples_order: 'RawSqlCommand' });
    expect(mockSqlClient.raw).toHaveBeenCalledWith('samples_order || \'["mockSampleId", "id2"]\'::jsonb');
    expect(mockSqlClient.where).toHaveBeenCalledWith('id', 'mockExperimentId');
  });

  it('deleteSample works correctly', async () => {
    mockSqlClient.where.mockImplementationOnce(() => { Promise.resolve(); });
    mockSqlClient.raw.mockImplementationOnce(() => 'RawSqlCommand');

    await new Experiment().deleteSample(mockExperimentId, mockSampleId);

    expect(mockSqlClient.update).toHaveBeenCalledWith({ samples_order: 'RawSqlCommand' });
    expect(mockSqlClient.raw).toHaveBeenCalledWith('samples_order - \'mockSampleId\'');
    expect(mockSqlClient.where).toHaveBeenCalledWith('id', 'mockExperimentId');
  });

  it('getProcessingConfig works', async () => {
    const mockFind = jest.spyOn(BasicModel.prototype, 'find')
      .mockImplementationOnce(() => Promise.resolve(getProcessingConfigResponse));

    const result = await new Experiment().getProcessingConfig(mockExperimentId);
    expect(mockFind).toHaveBeenCalledWith({ id: mockExperimentId });
    expect(result).toMatchSnapshot();
  });

  it('getProcessingConfig throws an error if there is no matching experiment', async () => {
    jest.spyOn(BasicModel.prototype, 'findOne')
      .mockImplementationOnce(() => Promise.resolve(undefined));

    expect(async () => {
      await new Experiment().getProcessingConfig(mockExperimentId);
    }).rejects.toThrow();
  });

  it('updateProcessingConfig works', async () => {
    const mockBody = [{
      name: 'classifier',
      body: {
        changedField: 'IamChanging so much',
      },
    }];

    mockSqlClient.where.mockImplementationOnce(() => { Promise.resolve(); });
    await new Experiment().updateProcessingConfig(mockExperimentId, mockBody);
    expect(mockSqlClient.where).toHaveBeenCalledWith('id', mockExperimentId);
  });

  it('downloadData returns signed url when the correct download type is given', async () => {
    const signedUrlSpy = mockS3GetSignedUrl();

    const experimentId = 'someExperiment-UUID-with-several-parts';
    const filenamePrefix = experimentId.split('-')[0];
    const expectedFileName = `${filenamePrefix}_processed_matrix.rds`;

    await new Experiment().getDownloadLink(experimentId, 'processed-matrix');

    expect(signedUrlSpy).toHaveBeenCalledWith(
      'getObject',
      {
        Bucket: `processed-matrix-test-${config.awsAccountId}`,
        Expires: 120,
        Key: `${experimentId}/r.rds`,
        ResponseContentDisposition: `attachment; filename ="${expectedFileName}"`,
      },
    );
  });

  it('downloadData throws error incorrect download type is given', async () => {
    new Experiment().getDownloadLink('12345', 'invalid type')
      .catch((error) => {
        expect(error.message).toMatch(/Invalid download type requested/gi);
      });
  });
});
