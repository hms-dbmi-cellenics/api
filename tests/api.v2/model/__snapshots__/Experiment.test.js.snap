// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model/Experiment getAllExperiments works correctly 1`] = `
Array [
  Array [
    [MockFunction],
    Array [
      "id",
      "name",
      "description",
      "samples_order",
      "notify_by_email",
      "pipeline_version",
      "created_at",
      "updated_at",
      "parent_experiment_id",
      "is_subsetted",
    ],
    "key",
    "metadataKeys",
    [MockFunction],
  ],
]
`;

exports[`model/Experiment getExperimentData works correctly 1`] = `
Array [
  "COALESCE(
      jsonb_object_agg(pipeline_type, jsonb_build_object('state_machine_arn', state_machine_arn, 'execution_arn', execution_arn))
      FILTER(
        WHERE pipeline_type IS NOT NULL
      ),
      '{}'::jsonb
    ) as pipelines,}), as pipelines",
]
`;

exports[`model/Experiment getProcessingConfig works 1`] = `
Object {
  "cellSizeDistribution": Object {
    "auto": true,
    "enabled": false,
    "filterSettings": Object {
      "binStep": 200,
      "minCellSize": 1080,
    },
    "sample1Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 200,
        "minCellSize": 907,
      },
      "enabled": false,
      "filterSettings": Object {
        "binStep": 200,
        "minCellSize": 907,
      },
    },
    "sample2Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 200,
        "minCellSize": 623,
      },
      "enabled": false,
      "filterSettings": Object {
        "binStep": 200,
        "minCellSize": 623,
      },
    },
    "sample3Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 200,
        "minCellSize": 516,
      },
      "enabled": false,
      "filterSettings": Object {
        "binStep": 200,
        "minCellSize": 516,
      },
    },
  },
  "classifier": Object {
    "brandNewTries": "changing lmaolxdd",
  },
  "configureEmbedding": Object {
    "apiUrl": "http://api.api-default.svc.cluster.local:3000",
    "clusteringSettings": Object {
      "method": "louvain",
      "methodSettings": Object {
        "louvain": Object {
          "resolution": 0.8,
        },
      },
    },
    "embeddingSettings": Object {
      "method": "umap",
      "methodSettings": Object {
        "tsne": Object {
          "learningRate": 938.9167,
          "perplexity": 30,
        },
        "umap": Object {
          "distanceMetric": "cosine",
          "minimumDistance": 0.3,
        },
      },
    },
  },
  "dataIntegration": Object {
    "apiUrl": "http://api.api-default.svc.cluster.local:3000",
    "dataIntegration": Object {
      "method": "harmony",
      "methodSettings": Object {
        "fastmnn": Object {
          "normalisation": "logNormalize",
          "numGenes": 2000,
        },
        "harmony": Object {
          "normalisation": "logNormalize",
          "numGenes": 2000,
        },
        "seuratv4": Object {
          "normalisation": "logNormalize",
          "numGenes": 2000,
        },
        "unisample": Object {
          "normalisation": "logNormalize",
          "numGenes": 2000,
        },
      },
    },
    "dimensionalityReduction": Object {
      "excludeGeneCategories": Array [],
      "method": "rpca",
      "numPCs": 30,
    },
  },
  "doubletScores": Object {
    "auto": true,
    "enabled": true,
    "filterSettings": Object {
      "binStep": 0.05,
      "probabilityThreshold": 0.5,
    },
    "sample1Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.3784758,
      },
      "enabled": true,
      "filterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.3784758,
      },
    },
    "sample2Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.4101394,
      },
      "enabled": true,
      "filterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.4101394,
      },
    },
    "sample3Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.7622447,
      },
      "enabled": true,
      "filterSettings": Object {
        "binStep": 0.05,
        "probabilityThreshold": 0.7622447,
      },
    },
  },
  "mitochondrialContent": Object {
    "auto": true,
    "enabled": true,
    "filterSettings": Object {
      "method": "absoluteThreshold",
      "methodSettings": Object {
        "absoluteThreshold": Object {
          "binStep": 0.05,
          "maxFraction": 0.1,
        },
      },
    },
    "sample1Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
    },
    "sample2Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
    },
    "sample3Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "method": "absoluteThreshold",
        "methodSettings": Object {
          "absoluteThreshold": Object {
            "binStep": 0.05,
            "maxFraction": 0.1,
          },
        },
      },
    },
  },
  "numGenesVsNumUmis": Object {
    "enabled": true,
    "filterSettings": Object {
      "regressionType": "gam",
      "regressionTypeSettings": Object {
        "gam": Object {
          "pLevel": 0.001,
        },
      },
    },
    "sample1Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009066183,
          },
          "spline": Object {
            "pLevel": 0.00009066183,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009066183,
          },
          "spline": Object {
            "pLevel": 0.00009066183,
          },
        },
      },
    },
    "sample2Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009075234,
          },
          "spline": Object {
            "pLevel": 0.00009075234,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009075234,
          },
          "spline": Object {
            "pLevel": 0.00009075234,
          },
        },
      },
    },
    "sample3Id": Object {
      "apiUrl": "http://api.api-default.svc.cluster.local:3000",
      "auto": true,
      "defaultFilterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009180207,
          },
          "spline": Object {
            "pLevel": 0.00009180207,
          },
        },
      },
      "enabled": true,
      "filterSettings": Object {
        "regressionType": "linear",
        "regressionTypeSettings": Object {
          "linear": Object {
            "pLevel": 0.00009180207,
          },
          "spline": Object {
            "pLevel": 0.00009180207,
          },
        },
      },
    },
  },
}
`;

exports[`model/Experiment updateSamplePosition rolls back if the parameters are invalid 1`] = `
Array [
  "(
        SELECT jsonb_insert(samples_order - 0, '{10000}', samples_order -> 0, false)
          FROM (
            SELECT (samples_order)
            FROM experiment e
            WHERE e.id = 'mockExperimentId'
          ) samples_order
        )",
]
`;

exports[`model/Experiment updateSamplePosition rolls back if the result is invalid 1`] = `
Array [
  "(
        SELECT jsonb_insert(samples_order - 0, '{1}', samples_order -> 0, false)
          FROM (
            SELECT (samples_order)
            FROM experiment e
            WHERE e.id = 'mockExperimentId'
          ) samples_order
        )",
]
`;

exports[`model/Experiment updateSamplePosition works correctly if valid params are passed 1`] = `
Array [
  "(
        SELECT jsonb_insert(samples_order - 0, '{1}', samples_order -> 0, false)
          FROM (
            SELECT (samples_order)
            FROM experiment e
            WHERE e.id = 'mockExperimentId'
          ) samples_order
        )",
]
`;
